fix(networking): Fix handling of incoming TLS connections

bruceg opened the pull request:
Closes #2123

Commit:
Restore "Use MaybeTlsListener in TcpSource"

This reverts commit bdedf7babd63a4724293ca6cba7503697ba65bb3.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move handling of TLS handshaking into new MaybeTlsStream

Incoming TLS handled handshaking inline with accepting a new connection.
However, this ended up leaving Vector ignoring new connections while
handshaking was being processed. By moving handshaking into the stream,
we can simply accept new connections into a MaybeTlsStream in the
"accepting" state, and deal with all of the I/O there and immediately be
ready for another connection.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Make some of the TcpStream specific code generic

Signed-off-by: Bruce Guenter <bruce@timber.io>

LucioFranco commented on src/sources/util/tcp.rs:
```diff
@@ -121,15 +135,8 @@ pub trait TcpSource: Clone + Send + 'static {
-                            out.clone(),
-                            tls.clone(),
-                        )
+                        debug!("accepted a new connection");
```
we could include the remote addr here.

bruceg replied:
Good idea. The old code didn't report the remote address neither, but it's still a good idea.

LucioFranco replied:
We might be able to append it to the span too via https://docs.rs/tracing/0.1.13/tracing/struct.Span.html#method.record

LucioFranco reviewed (approved):
LGTM!

Commit:
Log the new peer address after accepting TCP connection

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop the Result from MaybeTlsStream::peer_addr

Since we have cached the peer address and handled the error when
creating the MaybeTlsStream, this accessor can never fail.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop unnecessary Option from source address in TcpSource trait

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix feature build issues

So much for keeping all the feature gates in the top-level `tls/mod.rs`

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Split incoming-only handling from plain MaybeTlsStream

This allows all the TLS handling required for sources to be separated
into the `tls/incoming.rs` file again.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Reduce the states in MaybeTlsIncomingStream using MaybeTlsStream

Signed-off-by: Bruce Guenter <bruce@timber.io>

LucioFranco commented on src/tls/incoming.rs:
```diff
@@ -98,3 +105,101 @@ impl From<TcpListener> for MaybeTlsListener {
     }
 }
+
+pub struct MaybeTlsIncomingStream<S> {
```
Why is this called with `Incoming` on it? This should just be a `MaybeTlsStream`? It can store an Option<PeerAddr> to make it more flexible.

bruceg replied:
So I initially rewrote the existing trivial `MaybeTlsStream` to handle the accept handshake case. However, this ended up complicating the sources with a lot (a dozen?) of `#[cfg]` flags due to the `Accepting` state variant never being constructed when TLS sources are compiled out. So, I restored the original `MaybeTlsStream` which handles the both a raw stream and a successfully negotiated TLS stream, and split off this incoming stream, which handles the negotiation and then drops into a stream state. It reduced the number of lines of code, so I figured it was a win.

LucioFranco replied:
Is this an issue to do with feature flags? if so I see that as a reason to remove them. Yeah, just seems odd that we need two types for this. It should be quite simple :/

bruceg replied:
Even discounting the feature flags issue, splitting this reduced the lines of code a bit. The diff from before I added all the feature flag gates (so it nulls out adding and removing them):

```
 src/tls/incoming.rs | 123 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------
 src/tls/mod.rs      |  16 ++++++++--
 src/tls/outgoing.rs |   4 +--
 src/tls/stream.rs   | 135 -------------------------------------------------------------------------------------
 4 files changed, 130 insertions(+), 148 deletions(-)
```

LucioFranco replied:
Yeah, mostly just find it confusing since for me my mental model is tcpstream wrapped by maybetlstream but with the incoming part it gets a bit fuzzy! but as long as it works I am happy.

bruceg commented on src/sources/socket/tcp.rs:
```diff
@@ -62,9 +62,7 @@ impl TcpSource for RawTcpSource {
-        if let Some(host) = host {
-            event.as_mut_log().insert(host_key.clone(), host);
-        }
+        event.as_mut_log().insert(host_key.clone(), host);
```
This is just propagating the now effectively unconditional `peer_addr` down into callers. I'm not clear what should be optional here.

LucioFranco replied:
Lets say you are proxing requests through something that would be the peer addr right? That isn't 100% the host that the log originated from and that could be confusing.

LucioFranco reviewed (approved):
Happy to move forward and get this overwith.
